typescript/phoenix-sdk/src/utils/market.ts

```diff
@@ -437,6 +478,134 @@ export function printUiLadder(uiLadder: UiLadder) {
+  for (const [orderId, restingOrder] of marketData.bids) {
+    if (restingOrder.lastValidSlot != 0 && restingOrder.lastValidSlot < slot) {
+      continue;
+    }
+    if (
+      restingOrder.lastValidUnixTimestampInSeconds != 0 &&
+      restingOrder.lastValidUnixTimestampInSeconds < unixTimestamp
+    ) {
+      continue;
+    }
+    const priceInTicks = toBN(orderId.priceInTicks);
+    const numBaseLots = toBN(restingOrder.numBaseLots);
+
+    bids.push({
+      priceInTicks,
+      sizeInBaseLots: numBaseLots,
+      side: Side.Bid,
+      makerPubkey: marketData.traderIndexToTraderPubkey.get(
+        toNum(restingOrder.traderIndex)
+      ),
+      orderSequenceNumber: getUiOrderSequenceNumber(orderId),
+    });
+
+    if (bids.length === ordersPerSide) {
+      break;
+    }
+  }
+
+  for (const [orderId, restingOrder] of marketData.asks) {
+    if (restingOrder.lastValidSlot != 0 && restingOrder.lastValidSlot < slot) {
+      continue;
+    }
+    if (
+      restingOrder.lastValidUnixTimestampInSeconds != 0 &&
+      restingOrder.lastValidUnixTimestampInSeconds < unixTimestamp
+    ) {
+      continue;
+    }
+    const priceInTicks = toBN(orderId.priceInTicks);
+    const numBaseLots = toBN(restingOrder.numBaseLots);
+
+    asks.push({
+      priceInTicks,
+      sizeInBaseLots: numBaseLots,
+      side: Side.Ask,
+      makerPubkey: marketData.traderIndexToTraderPubkey.get(
+        toNum(restingOrder.traderIndex)
+      ),
+      orderSequenceNumber: getUiOrderSequenceNumber(orderId),
+    });
+
+    if (asks.length === ordersPerSide) {
+      break;
+    }
+  }
+
+  return {
+    asks,
+    bids,
+  };
+}
+
+/**
+ * Returns the L3 book of bids and asks as JS numbers
+ *
+ * @param marketData The `MarketData` to get the ladder from
+ * @param levels The number of book levels to return
+ */
+export function getMarketL3UiBook(
+  marketData: MarketData,
+  ordersPerSide: number = DEFAULT_L3_BOOK_DEPTH,
+  slot: beet.bignum = 0,
+  unixTimestamp: beet.bignum = 0
+): L3UiBook {
+  const l3Book = getMarketL3Book(
+    marketData,
+    slot,
+    unixTimestamp,
+    ordersPerSide
+  );
+
+  return {
+    bids: l3Book.bids.map((b) => getL3UiOrder(b, marketData)),
+    asks: l3Book.asks.map((a) => getL3UiOrder(a, marketData)),
+  };
+}
+
+/**
+ * Converts a L3 order from BN to JS number representation
+ *
+ * @param l3Order The L3 order to convert
+ * @param marketData The `MarketData` the order was taken from
+ */
+function getL3UiOrder(l3Order: L3Order, marketData: MarketData): L3UiOrder {
+  return {
+    price:
+      (toNum(l3Order.priceInTicks) *
+        marketData.quoteLotsPerBaseUnitPerTick *
+        toNum(marketData.header.quoteLotSize)) /
+      (10 ** marketData.header.quoteParams.decimals *
+        marketData.header.rawBaseUnitsPerBaseUnit),
+    side: l3Order.side,
+    size:
+      (toNum(l3Order.sizeInBaseLots) *
+        marketData.header.rawBaseUnitsPerBaseUnit) /
+      marketData.baseLotsPerBaseUnit,
```

Comment:
this is kind of cancer. not sure if these methods are in the right place, will audit later

Reply:
I think this is right, but we should maybe use some canonical conversions helpers to make this stuff consistent.

Reply:
Right now we have a bunch of conversion helpers on the Client, maybe these should live on the Market or just in isolation